### ISSUE TITLE
Panic with multiple yields

### ISSUE THREAD

**Issue by nathanielc** [influxdata/flux#4780]
This query panics with multiple yields:

```
import "contrib/anaisdg/statsmodels"

from(bucket: "longterm")
  |> range(start: v.timeRangeStart, stop: v.timeRangeStop)
  |> filter(fn: (r) => r._measurement == "shift")
  |> filter(fn: (r) => r._field == "total" and not exists r.condition)
  |> yield(name: "context")
  |> toFloat()
  |> statsmodels.linearRegression()
  |> map(fn: (r) => ({ r with _value: r.y_hat }))
  |> yield(name: "trend")
```

```
panic during program start" log_id=0aUo7qcl000 service=storage-reads error="runtime error: invalid memory address or nil pointer dereference" stacktrace="goroutine 2139 [running]:
runtime/debug.Stack()
	/go/src/runtime/debug/stack.go:24 +0x88
github.com/influxdata/influxdb/v2/query/control.(*Controller).executeQuery.func1(0x4007226900, 0x400a97d440)
	/root/project/query/control/controller.go:465 +0x140
panic({0xffffa6be38e0, 0xffffa8169a70})
	/go/src/runtime/panic.go:1038 +0x21c
github.com/influxdata/flux/interpreter.(*Interpreter).doExpression(0x400959f180, {0xffffa6fbb4e0, 0x4017768630}, {0xffffa6fcc0d8, 0x4000fd3dd0}, {0xffffa6ff3148, 0x4017768a50})
	/go/pkg/mod/github.com/influxdata/flux@v0.161.0/interpreter/interpreter.go:374 +0x1550
github.com/influxdata/flux/interpreter.(*Interpreter).doExpression(0x400959f180, {0xffffa6fbb4e0, 0x4017768630}, {0xffffa6fcbce8, 0x4000fd3d40}, {0xffffa6ff3148, 0x4017768a50})
	/go/pkg/mod/github.com/influxdata/flux@v0.161.0/interpreter/interpreter.go:428 +0x1114
github.com/influxdata/flux/interpreter.(*Interpreter).doVariableAssignment(0x400959f180, {0xffffa6fbb4e0, 0x4017768630}, 0x40010dacc0, {0xffffa6ff3148, 0x4017768a50})
	/go/pkg/mod/github.com/influxdata/flux@v0.161.0/interpreter/interpreter.go:319 +0x5c
github.com/influxdata/flux/interpreter.(*Interpreter).doStatement(0x400959f180, {0xffffa6fbb4e0, 0x4017768630}, {0xffffa6fc67b8, 0x40010dacc0}, {0xffffa6ff3148, 0x4017768a50})
	/go/pkg/mod/github.com/influxdata/flux@v0.161.0/interpreter/interpreter.go:199 +0x3cc
github.com/influxdata/flux/interpreter.function.doCall({0x400104ed00, {0xffffa6ff3148, 0x400fe4d470}, 0x400959f180}, {0xffffa6fbb4e0, 0x4017768630}, {0xffffa7004c18, 0x400eb6a3f0})
	/go/pkg/mod/github.com/influxdata/flux@v0.161.0/interpreter/interpreter.go:938 +0x1e0
github.com/influxdata/flux/interpreter.function.Call({0x400104ed00, {0xffffa6ff3148, 0x400fe4d470}, 0x400959f180}, {0xffffa6fbb4e0, 0x4017768630}, {0xffffa7004178, 0x400bd7c460})
	/go/pkg/mod/github.com/influxdata/flux@v0.161.0/interpreter/interpreter.go:886 +0x108
github.com/influxdata/flux/interpreter.(*Interpreter).doCall(0x400959f180, {0xffffa6fbb4e0, 0x400a6d08d0}, 0x400b3bf050, {0xffffa6ff3148, 0x400fe4c990})
	/go/pkg/mod/github.com/influxdata/flux@v0.161.0/interpreter/interpreter.go:724 +0x474
github.com/influxdata/flux/interpreter.(*Interpreter).doExpression(0x400959f180, {0xffffa6fbb4e0, 0x400a6d08d0}, {0xffffa6fcbd78, 0x400b3bf050}, {0xffffa6ff3148, 0x400fe4c990})
	/go/pkg/mod/github.com/influxdata/flux@v0.161.0/interpreter/interpreter.go:368 +0x13f4
github.com/influxdata/flux/interpreter.(*Interpreter).doArguments.func1(0x400eb6a258)
	/go/pkg/mod/github.com/influxdata/flux@v0.161.0/interpreter/interpreter.go:788 +0x31c
github.com/influxdata/flux/values.BuildObjectWithSize(0x0, 0x4003a271b0)
	/go/pkg/mod/github.com/influxdata/flux@v0.161.0/values/object.go:113 +0x164
github.com/influxdata/flux/values.BuildObject(...)
	/go/pkg/mod/github.com/influxdata/flux@v0.161.0/values/object.go:100
github.com/influxdata/flux/interpreter.(*Interpreter).doArguments(0x400959f180, {0xffffa6fbb4e0, 0x400a6d08d0}, 0x400935a680, {0xffffa6ff3148, 0x400fe4c990}, {0x5, {0xffffa6f923e8, 0x400086a7a0}, 0xc}, ...)
	/go/pkg/mod/github.com/influxdata/flux@v0.161.0/interpreter/interpreter.go:775 +0x340
github.com/influxdata/flux/interpreter.(*Interpreter).doCall(0x400959f180, {0xffffa6fbb4e0, 0x400a6d08d0}, 0x400b3befc0, {0xffffa6ff3148, 0x400fe4c990})
	/go/pkg/mod/github.com/influxdata/flux@v0.161.0/interpreter/interpreter.go:702 +0x288
github.com/influxdata/flux/interpreter.(*Interpreter).doExpression(0x400959f180, {0xffffa6fbb4e0, 0x400a6d08d0}, {0xffffa6fcbd78, 0x400b3befc0}, {0xffffa6ff3148, 0x400fe4c990})
	/go/pkg/mod/github.com/influxdata/flux@v0.161.0/interpreter/interpreter.go:368 +0x13f4
github.com/influxdata/flux/interpreter.(*Interpreter).doStatement(0x400959f180, {0xffffa6fbb4e0, 0x400a6d08d0}, {0xffffa6fc66b8, 0x40088e2870}, {0xffffa6ff3148, 0x400fe4c990})
	/go/pkg/mod/github.com/influxdata/flux@v0.161.0/interpreter/interpreter.go:203 +0x26c
github.com/influxdata/flux/interpreter.(*Interpreter).doFile(0x400959f180, {0xffffa6fbb4e0, 0x400a6d08d0}, 0x4009437980, {0xffffa6ff3148, 0x400fe4c990}, {0xffffa6f797e0, 0x400915dee0})
	/go/pkg/mod/github.com/influxdata/flux@v0.161.0/interpreter/interpreter.go:111 +0x184
github.com/influxdata/flux/interpreter.(*Interpreter).doPackage(0x400959f180, {0xffffa6fbb4e0, 0x400a6d08d0}, 0x400419f180, {0xffffa6ff3148, 0x400fe4c990}, {0xffffa6f797e0, 0x400915dee0})
	/go/pkg/mod/github.com/influxdata/flux@v0.161.0/interpreter/interpreter.go:94 +0x98
github.com/influxdata/flux/interpreter.(*Interpreter).doRoot(0x400959f180, {0xffffa6fbb4e0, 0x400a6d08d0}, {0xffffa6fbc9a8, 0x400419f180}, {0xffffa6ff3148, 0x400fe4c990}, {0xffffa6f797e0, 0x400915dee0})
	/go/pkg/mod/github.com/influxdata/flux@v0.161.0/interpreter/interpreter.go:84 +0x114
github.com/influxdata/flux/interpreter.(*Interpreter).Eval(0x400959f180, {0xffffa6fbb4e0, 0x400a6d08d0}, {0xffffa6fbc9a8, 0x400419f180}, {0xffffa6ff3148, 0x400fe4c990}, {0xffffa6f797e0, 0x400915dee0})
	/go/pkg/mod/github.com/influxdata/flux@v0.161.0/interpreter/interpreter.go:75 +0x6c
github.com/influxdata/flux/runtime.(*runtime).Eval(0xffffadb07fb0, {0xffffa6fbb4e0, 0x400a6d08d0}, {0xffffa6f9b6e8, 0x4004f910a0}, {0xffffa6f92578, 0xffffadb415a0}, {0x400915de00, 0x2, 0x2})
	/go/pkg/mod/github.com/influxdata/flux@v0.161.0/runtime/runtime.go:119 +0x21c
github.com/influxdata/flux/lang.(*AstProgram).getSpec(0x40088e2280, {0xffffa6fbb4e0, 0x400a6d08a0}, {0xffffa6fbc350, 0x40088e2370})
	/go/pkg/mod/github.com/influxdata/flux@v0.161.0/lang/compiler.go:449 +0x224
github.com/influxdata/flux/lang.(*AstProgram).Start(0x40088e2280, {0xffffa6fbb4e0, 0x400a6d06f0}, {0xffffa6fbc350, 0x40088e2370})
	/go/pkg/mod/github.com/influxdata/flux@v0.161.0/lang/compiler.go:492 +0x200
github.com/influxdata/influxdb/v2/query/control.(*Controller).executeQuery(0x400a97d440, 0x4007226900)
	/root/project/query/control/controller.go:488 +0x210
github.com/influxdata/influxdb/v2/query/control.(*Controller).processQueryQueue(0x400a97d440)
	/root/project/query/control/controller.go:446 +0x34
github.com/influxdata/influxdb/v2/query/control.New.func1(0x400a97d440)
	/root/project/query/control/controller.go:231 +0x54
created by github.com/influxdata/influxdb/v2/query/control.New
	/root/project/query/control/controller.go:229 +0x844
```

This was discovered in OSS full detail here https://community.influxdata.com/t/internal-error-when-using-statsmodels-linearregression/25055

**Comment by onelson** [comment]
Trying to reduce the repro a little bit. Some initial observations follow.

This panics:
```
sampledata.int()
  |> yield(name: "early")
  |> toFloat()
  |> statsmodels.linearRegression()
```
Removing the early yield makes the query work.


Similarly, this returns an error:
```
sampledata.float()
  |> yield(name: "early")
  |> statsmodels.linearRegression()
```

`Error: error calling function "linearRegression" @7:6-7:36: unsupported binary expression invalid / invalid`

Removing the early yield allows the query to work.

---

The through-line being having a yield as a part of the input tables for `linearRegression()`, not the number of yields.

Still unsure as to what's going on, but I'd like to dig in on the panic first. Examining the hard failure may give context for the error in the second case, also possibly a common cause.

I'll note that `linearRegression()` relies on `tableFind()` which has recently been discussed as having quirks when it comes to yields, ref: #4805

**Comment by onelson** [comment]
# Recap

To summarize the above investigation, we've got 2 problems surfaced by the repro:

- tables can be marked as `Done` too early when `yield` and `tableFind` are used together.
- When result count is >1 `tableFind` will only return one (random) result from its input.

#4813 will address the latter. I'm going to use this issue to try and correct the former.

**Comment by onelson** [comment]
I've got a (one line) patch that appears to fix the repro case, but I'm struggling to write tests in support of the change.

In order to write a flux acceptance test, based on the repro, which uses a `yield` upstream of a `tableFind`, I think we would first need to implement #4300. This is to say, having a non-terminal non-empty yield in the testcase is an automatic failure, but it's a requirement for this repro.

In order to write a more low-level unit test, I'm finding I need to understand more of the Table API. Trying to fumble through by mimicking existing tests gave bad outcomes. Going to keep investigating this.

**Comment by onelson** [comment]
With #4820 merged, I'm moving on to #4813 which I believe should complete this issue.

**Comment by onelson** [comment]
With the fixes in #4869 for #4813, this should work as expected.
